pkg/chainreader/contract_reader.go

```diff
@@ -0,0 +1,111 @@
+package chainreader
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+)
+
+// ContractReaderByIDs wraps types.ContractReader to allow the caller to set custom contractIDs with Bind.
+type ContractReaderByIDs interface {
+	// Bind accepts a map of bound contracts where map keys are custom contractIDs to be used for calling the interface.
+	Bind(ctx context.Context, bindings map[string]types.BoundContract) error
+	Unbind(ctx context.Context, bindings map[string]types.BoundContract) error
+	GetLatestValue(ctx context.Context, contractID, readName string, confidenceLevel primitives.ConfidenceLevel, params, returnVal any) error
+	BatchGetLatestValues(ctx context.Context, request BatchGetLatestValuesRequestByCustomID) (BatchGetLatestValuesResultByCustomID, error)
+	QueryKey(ctx context.Context, contractID string, filter query.KeyFilter, limitAndSort query.LimitAndSort, sequenceDataType any) ([]types.Sequence, error)
+}
+
+// WrapContractReaderByIDs returns types.ContractReader behind ContractReaderByIDs interface.
+func WrapContractReaderByIDs(contractReader types.ContractReader) ContractReaderByIDs {
+	return &contractReaderByIDs{
+		cr: contractReader,
+	}
+}
+
+type contractReaderByIDs struct {
+	bindings sync.Map
+	cr       types.ContractReader
+}
```

Comment:
Does this need to embed `UnimplementedContractReader`?

Reply:
Not really, `UnimplementedContractReader` is mainly for relayer impl. to have a stub so that the build doesn't break. If there is an interface change to Contract Reader, chainlink-common won't be mergeable before `contractReaderByIDs` calls to ContractReader are updated. There is a separate concern here though, if the interface changes are breaking then the wrapper methods will have to be modified too, which means that we would have to version both the ContractReader interface and this wrapper.